libraries/chain/apply_context.cpp

```diff
@@ -120,12 +120,44 @@ void apply_context::exec_one()
    //    * a pointer to an object in a chainbase index is not invalidated if other objects in that index are modified, removed, or added;
    //    * a pointer to an object in a chainbase index is not invalidated if the fields of that object are modified;
    //    * and, the *receiver_account object itself cannot be removed because accounts cannot be deleted in EOSIO.
+   const auto& encode = [&](char* bytes, uint32_t len) {
+      const size_t bs = eosio::chain::config::hashing_checktime_block_size;
+      sha256::encoder enc;
+      while ( len > bs ) {
+         enc.write( bytes, bs );
+         bytes += bs;
+         len  -= bs;
+         trx_context.checktime();
+      }
+      enc.write( bytes, len );
+      return enc.result();
+   };
+
+   const auto& encode_action = [&](const action& act, action_receipt& r) {
+      if (control.is_builtin_activated( builtin_protocol_feature_t::action_return_value )) {
+         const action_base* base = &act; // downcast to get only the action_base packed size
+         const auto act_ret_val_size = fc::raw::pack_size(action_return_value) + 1; // + 1 for optional
+         const auto act_pack_size    = fc::raw::pack_size(act);
+         std::vector<char> buff(act_pack_size + act_ret_val_size);
+         datastream<char*> ds(buff.data(), act_pack_size + act_ret_val_size);
+         fc::raw::pack(ds, act);
+         fc::raw::pack(ds, (uint8_t)1);
+         fc::raw::pack(ds, action_return_value);
+         sha256 hashes[2];
+         const size_t lhs_size = fc::raw::pack_size(*base);
+         hashes[0] = encode(buff.data(), lhs_size);
+         hashes[1] = encode(buff.data() + lhs_size, fc::raw::pack_size(act.data) + act_ret_val_size);
+         r.act_digest = encode((char*)&hashes[0], sizeof(hashes));
+      } else {
+         r.act_digest = digest_type::hash(act);
+      }
+   };
```

Comment:
I think this should be lifted out of here and put somewhere else. Maybe `action_receipt`. Then it could be tested instead duplicating the code in the tests. `encode` could be passed in so tests could use a non-checktime version.

Reply:
Sure, I waffled on this a lot and went with the path of least impact first.  And then made a bunch of changes elsewhere that negated that small impact, so will do make it a method of action_receipt and have it parameterized to either checktime or not.